arcconfig/arcconfig.go

```diff
@@ -0,0 +1,164 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package arcconfig defines the API Runtime Control config and provides
+// the config loading and validation functions.
+package arcconfig
+
+import "time"
+
+// APIRuntimeControlConfig is an Apigee Environment-level config for
+// Envoy Adapter. It contains a list of operations for the adapter to
+// perform request authentication and authorization.
+type APIRuntimeControlConfig struct {
```

Comment:
We're missing a layer of indirection, we need the top level (the Environment), and then another layer for each Proxy (i.e. per base path) so something like:

```
type EnvironmentConfig struct {
  ID string
  []ProxyConfig ProxyConfigs
}

type ProxyConfig {
  BasePath string
  []Operation Operations 
  ...
}
```

Reply:
Yup. I just asked about this as well. We do need to tie these to an environment somehow.

Reply:
re: Scott - I just thought we still need the original config file anyway which indicates the environment. As such, we only need a list of these configs assuming they are all tied to the specified environment. Or are we ditching the original config entirely if we see any of these ones?

Will add this layer. Apart from this, my idea was to have each operation self-contained here instead of having those global settings. I don't have strong opinion. If the goal is to make the structures here as alike the schema as possible, I will add those global rules back.

Reply:
Yeah, this is a little tricky.... Right now we have a global config element for env. And we have the ability to set the env from Envoy via metadata. We could keep using that scheme even with this config file. But only if these configs are not required to be tied to a specific env?

Reply:
So the requirement here is simply that there is some unique identifier per config (the `ID` field) that is linked to a complete group of Proxy configurations (basepath + operations). I don't fully understand how this will fit into the other "global" element... maybe a small YAML example would clarify?

Reply:
Actually reflecting on this, I recall a discussion that the global config would contain *references* to other files, where each file was a single `EnvironmentConfig` (i.e. group of Proxies).

Reply:
In case it helps, the existing "global" config looks like this:
```
global:
  temp_dir: /tmp/apigee-istio
  keep_alive_max_connection_age: 10m
  api_address: :5000
  metrics_address: :5001
tenant:
  remote_service_api: https://org-test.apigee.net/remote-service
  org_name: org
  env_name: env
  key: mykey
  secret: mysecret
  client_timeout: 30s
products:
  refresh_rate: 2m
analytics:
  file_limit: 1024
  send_channel_size: 10
  collection_interval: 10s
auth:
  api_key_claim: claim
  api_key_cache_duration: 30m
```
My understanding is if the configs defined here are specified, then the `auth` section in the original one will be ignored.

Reply:
Can we add a new field? e.g. 

```
arc_env_configs: ["file1", "file2"]
```

This would be mutually exclusive with `auth`

Reply:
The idea was to take these new configs via a new command line flag. But this new field makes sense to me as it gives the feasibility to still keep everything in one file.

Reply:
Yes, we could add a section like this. My original thought was that we'd simply accept the arc configs as a CLI params... but this would be more formal.

Reply:
It seems we all agree on such a new section then. Should I move everything into the `server` pkg then?

Reply:
Perhaps we should relocate all the config structs and processing to a config package.

Reply:
That sounds better to me. It wouldn't break any user-facing interfaces.